### DISCUSSION TITLE
Object must implement IConvertible requirement?

### DISCUSSION THREAD

**Discussion by e-mair** [sebastienros/jint#1128] [Q&A]
Hi,

I am facing an issue regarding the use of generics. I essentially try to pass an array of items to a function via `Engine.Invoke`, which in turn, gives me a `Object must implement IConvertible` error. Had a look through issues and found #15 - IConvertible requirement, not sure if this is related?

I have a script that accepts an array of people, that could look something like this:
```
let ipsum = (function(people) {
    return people.map(p => {
        return {
            FirstName: p.FirstName,
            Surname: p.LastName,
            FullName: p.FirstName + " " + p.MiddleName + " " + p.LastName,
            EmployeeNumber: p.MainEmployment.EmpNo
        }
    });
});
```

I have written an engine wrapper in C#, of which, has a ModifyItem<T> method with the following body:
```
public BusinessResult<object> ModifyItem<T>(OCSDataProfileConfigurationSVC config, T item)
{
     try
     {
          var safeProfileName = GenerateRandomVariableName();
          var safeCollectionName = GenerateRandomVariableName();
          var jsFunction = Engine.Execute($"let {safeProfileName} = {config.Script}").Evaluate(safeProfileName);
          var modifiedItem = ((IList<object>)Engine.Invoke(jsFunction, new T[] { item }).ToObject()).FirstOrDefault();

          if (modifiedItem is null)
               throw new ArgumentNullException(nameof(modifiedItem));

          return BusinessResult.Ok(modifiedItem);
     }
     catch (Exception e)
     {
          _logger.LogError("Failed to modify item of {type}: {errorMessage}", typeof(T).Name, e.Message);
          return BusinessResult.Fail<object>(e);
     }
}
```
Notice that I am initializing a new array of T when I call `Engine.Invoke` - we do this so that our scripts are generalized; we're essentially fooling the script into handling a collection of items when there will always be one item.

Any help would be appreciated! 🙏🏻

**Comment by lahma** [comment]
Could you please formulate a smallest possible failing test case (xUnit, NUnit, console program) for this problem? It would make investigating easier.

**Comment by e-mair** [comment]
## Dependencies
#### Person Class
```
public class Person
{
      public string FirstName { get; set; }

      public string LastName { get; set; }

     public long EmpNo { get; set; }

     public long CompanyId { get; set; }
}
```
This is the object I am trying to pass along to my JavaScript function. 
#### JavaScript
```
(function(people) {
    return people.map(p => {
        return {
            FirstName: p.FirstName,
            Surname: p.LastName,
            FullName: p.FirstName + " " + p.LastName,
            EmployeeNumber: p.EmpNo
        }
    });
});
```
This is the simple function we're attempting to call. It is designed to support a collection of people that need to be processed via `.map()`. 

I'll write a brief unit test later and post back what I find (if anything). Ideally I'd avoid having to implement `IConvertible` on my classes as I don't want to interfere with serialization.
